### DISCUSSION TITLE
Frequency graphing Math function feasibility?

### DISCUSSION THREAD

**Discussion by Agoutii** [OpenHantek/OpenHantek6022#397] [Ideas]
Hi all,

I'm looking for a USB oscilloscope package that can graph the frequency of a square wave in a capture, the frequency modulation you could say. The only package that I know does this is the Picoscope Automotive, which is horrendously expensive for a home user.

I believe the way it works is the waveform is auto-triggered as it would be to take a single frequency measurement, except the period inverse between every trigger is then output (or graphed directly) as the peak to peak frequency. It is also output as a live number in auto-trigger.

The use cases for this are in automotive testing. One is where you have frequency modulated signals (like digital MAF sensors) for which you have to monitor the frequency while changing engine throttle and rpm. Another is measuring the engine RPM per pulse of the crank sensor to identify things like misfires or to help narrow down bad compression on cylinders.

I've included some video examples to help illustrate what I'm talking about - these are just the first ones I found through Google, no affiliation.

https://youtu.be/UUklzebyuyo?t=198
https://youtu.be/WnV9QsGreHc?t=198

Coincidentally the same time for both.

Is this something that is feasible to achieve with OpenHantek? I'm happy to learn and contribute however would be most appropriate - I'm not just demanding someone else do it - but I also don't have a good understanding of how hard this would be to implement.

**Comment by diogenescreosote** [comment]
The math itself shouldn't be hard. You're describing a classic case of the [Fast Fourier Transform](https://en.wikipedia.org/wiki/Fast_Fourier_transform), for which well-vetted open source libraries exist.

I agree that it would be useful to visualize signals in the frequency domain. Until someone implements it, you can do something similar as follows:

- Record the signal of interest as CSV
- Import the CSV into Excel or LibreOffice
   - If in Excel, do [this](https://www.ehow.co.uk/how_4490433_calculate-volume-sphere-microsoft-excel.html/) 
   - If in LibreOffice, do [this](https://dennisfrancis.wordpress.com/2019/03/27/discrete-fourier-transform-in-calc/)
- Plot the frequency series. This will show intensity as a function of frequency, for the **entire** duration of your data. It will not show intensity of a **given** frequency as a function of time, as shown in the videos. Still, it might give you the information you need.

Those videos seem to plot the intensity of a given frequency range over time (380Hz-440Hz at one point). Doing that requires some more specific jujitsu, which is quite doable in R/Matlab/Python, but not as as straightforward in spreadsheets.
